src/Features/Core/Portable/Testing/AbstractTestMethodFinder.cs

```diff
@@ -0,0 +1,107 @@
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.Testing;
+
+internal abstract class AbstractTestMethodFinder<TMethodDeclaration>(IEnumerable<ITestFrameworkMetadata> testFrameworks) : ITestMethodFinder where TMethodDeclaration : SyntaxNode
+{
+    /// <summary>
+    /// Output the method symbol as a fully qualified method name, e.g. Namespace.Class.Method to match what test discovery gives us.
+    /// Generics are not applicable here - none of our supported test frameworks allow generic test classes / methods.
+    /// </summary>
+    private static readonly SymbolDisplayFormat s_methodSymbolNoParametersDisplayFormat = new(
+        typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+        memberOptions: SymbolDisplayMemberOptions.IncludeContainingType);
+
+    protected readonly ImmutableArray<ITestFrameworkMetadata> TestFrameworkMetadata = testFrameworks.ToImmutableArray();
+
+    protected abstract bool IsTestMethod(TMethodDeclaration method);
+
+    protected abstract bool DescendIntoChildren(SyntaxNode node);
+
+    public async Task<ImmutableArray<SyntaxNode>> GetPotentialTestMethodsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        var testNodes = await GetPotentialTestNodesAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+
+        // Find any test methods that intersect with the requested span.
+        var intersectingNodes = testNodes.Where(node => node.Span.IntersectsWith(textSpan)).ToImmutableArray();
+        if (!intersectingNodes.IsEmpty)
+        {
+            return intersectingNodes;
+        }
+
+        // We might have been invoked on a test class.  Check if any of the test method parent nodes intersect with the requested text span.
+        return testNodes.Where(node => node.Parent?.Span.IntersectsWith(textSpan) == true).ToImmutableArray();
+    }
+
+    public bool IsMatch(SemanticModel semanticModel, SyntaxNode node, string fullyQualifiedTestName, CancellationToken cancellationToken)
+    {
+        var method = node as TMethodDeclaration;
+        Contract.ThrowIfNull(method, $"Node with kind {node.RawKind} is not a method");
+
+        // Since discovered tests are not guarantied to run on a particular snapshot, we match optimistically based on test name.
+        var methodSymbol = semanticModel.GetDeclaredSymbol(method, cancellationToken);
+        Contract.ThrowIfNull(methodSymbol, "Test method has no symbol");
+
+        var fullyQualifiedMethodName = methodSymbol.ToDisplayString(s_methodSymbolNoParametersDisplayFormat);
```

Comment:
is there any sort of up front check we can do on the name first, to avoid ToDisplayString.  It's *not* cheap.

Reply:
Hmm.  I could potentially check if the method syntax name is part of the fullyQualified name, and bail out if not.